src/NuGet.Clients/NuGet.CommandLine/MsBuildUtility.cs

```diff
@@ -98,155 +99,34 @@ public static bool IsMsBuildBasedProject(string projectFullPath)
-
-                // Disallow the import of targets/props from packages
-                argumentBuilder.Append(" /p:ExcludeRestorePackageImports=true ");
-
-                if (!string.IsNullOrEmpty(solutionDirectory))
-                {
-                    argumentBuilder.Append(" /p:RestoreSolutionDirectory=");
-                    argumentBuilder.Append(EscapeQuoted(solutionDirectory));
-                }
-
-                if (!string.IsNullOrEmpty(restoreConfigFile))
-                {
-                    argumentBuilder.Append(" /p:RestoreConfigFile=");
-                    argumentBuilder.Append(EscapeQuoted(restoreConfigFile));
-                }
-
-                var isMono = RuntimeEnvironmentHelper.IsMono && !RuntimeEnvironmentHelper.IsWindows;
-
-                if (sources.Length != 0)
-                {
-                    if (isMono)
-                    {
-                        argumentBuilder.Append(" /p:RestoreSources=\\\"");
-                    }
-                    else
-                    {
-                        argumentBuilder.Append(" /p:RestoreSources=\"");
-                    }
-
-                    for (var i = 0; i < sources.Length; i++)
-                    {
-                        if (isMono)
-                        {
-                            argumentBuilder.Append(sources[i])
-                                .Append("\\;");
-                        }
-                        else
-                        {
-                            argumentBuilder.Append(sources[i])
-                                .Append(";");
-                        }
-                    }
-
-                    if (isMono)
-                    {
-                        argumentBuilder.Append("\\\" ");
-                    }
-                    else
-                    {
-                        argumentBuilder.Append("\" ");
-                    }
-                }
-                if (!string.IsNullOrEmpty(packagesDirectory))
+                // Build a .targets file of all restore inputs, this is needed to avoid going over the limit on command line arguments.
+                var properties = new Dictionary<string, string>()
                 {
-                    argumentBuilder.Append(" /p:RestorePackagesPath=");
-                    AppendQuoted(argumentBuilder, packagesDirectory);
-                }
+                    { "RestoreUseCustomAfterTargets", "true" },
+                    { "RestoreGraphOutputPath", resultsPath },
+                    { "RestoreRecursive", recursive.ToString().ToLowerInvariant() },
+                    { "RestoreProjectFilterMode", "exclusionlist" }
+                };
 
-                // Add all depenencies as top level restore projects if recursive is set
-                argumentBuilder.Append($" /p:RestoreRecursive={recursive} ");
+                var inputTargetXML = GetRestoreInputFile(entryPointTargetPath, properties, projectPaths);
```

Comment:
I thought we wanted to include the restore sources/packages path etc in the targets file too?